### PR TITLE
Optimize ctint configuration.

### PR DESCRIPTION
**Pull request by gbalduzz**
Apply optimizations similar to #160 to the ct-int configuration.

- Much saner interface to the configuration, designed just for the submatrix update.
- O(1) retrieval of the matrix indices and vertex removal.
- Avoid repeated allocation.

The access to a random partner in case of multiple updates is still not optimized (see comment in code), but it should happen rarely and only in certain applications and would require either a fancy data structure, or relying on the statistical uniformity of std::hash.

### CODE REVIEW

File: cmake/dca_compiler_options.cmake [hunk 1]
```diff
@@ -21,7 +21,7 @@
 set(DCA_WARNINGS -Wall -Wextra -Wpedantic -Wno-sign-compare -Wno-dangling-else)
 
 # Languange standard
-set(DCA_STD_FLAG -std=c++14)
+set(DCA_STD_FLAG -std=c++17)
```
Comment (gbalduzz) [inline cmake/dca_compiler_options.cmake]: as in #160: not strictly necessary, but it would be nice.

**Comment by uhaehner** [comment]
retest this please

File: cmake/dca_compiler_options.cmake (hunk 1 above)
Comment (PDoakORNL) [reply cmake/dca_compiler_options.cmake]: I think while you are here is a good time to make this transition.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp [hunk 2]
```diff
@@ -28,22 +28,44 @@ namespace ctint {
...
     return aux_spin == b.aux_spin && interaction_id == b.interaction_id && tau == b.tau;
   }
 };
 
+struct ConfigRef {
+  ConfigRef() = default;
+  ConfigRef(unsigned _config_id, std::uint8_t _leg_id) : config_id(_config_id), leg_id(_leg_id) {}
+
+  unsigned config_id;   // Index of the interaction in the SolverConfiguration.
+  std::uint8_t leg_id;  // Stores if this is the first or second leg of the interaction.
```
Comment (PDoakORNL) [inline include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp]: So are the only valid values 1 and 2 or is there more to it?

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp [hunk 3]
```diff
@@ -46,54 +49,29 @@ class SolverConfiguration : public MatrixConfiguration {
   template <class RngType>
   void insertRandom(RngType& rng);
 
-  // TODO: pass output as argument.
+  // Returns at most two candidates. -1 stands for a missing candidate.
```
Comment (PDoakORNL) [inline include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: well now it always returns an array of size 2.  So its not going to be -1, update the documentation please.
Documenta removal_rand.

File: src/phys/dca_step/cluster_solver/ctint/structs/ctint_matrix_configuration.cpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+namespace dca {
+namespace phys {
+namespace solver {
+namespace ctint {
+// dca::phys::solver::ctint::
+
+MatrixConfiguration::MatrixConfiguration(const InteractionVertices* H_int, const int bands)
+    : H_int_(H_int), n_bands_(bands), sectors_{Sector(), Sector()} {}
+
+MatrixConfiguration& MatrixConfiguration::operator=(const MatrixConfiguration& rhs) {
```
Comment (PDoakORNL) [inline src/phys/dca_step/cluster_solver/ctint/structs/ctint_matrix_configuration.cpp]: These don't follow the semantics I'd expect for an assignment operator. i.e. H_int_ and n_bands_ are not assigned. Either add those or just use a named function like updateSectors.

Comment (gbalduzz) [reply src/phys/dca_step/cluster_solver/ctint/structs/ctint_matrix_configuration.cpp]: updated in 727f88b2 with an assertion on the const member.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp (hunk 3 above)
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: Updated the comment. I want something that returns up to two indices, with no allocation, so I return -1 for a missing index.

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp (hunk 2 above)
Comment (gbalduzz) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/ct_int_matrix_configuration.hpp]: 0 and 1. Is the comment confusing? I updated it in 727f88b2

File: include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp (hunk 3 above)
Comment (PDoakORNL) [reply include/dca/phys/dca_step/cluster_solver/ctint/structs/solver_configuration.hpp]: Ah I see for either one.

**Pull request merged**